aptos-move/aptos-aggregator/src/aggregator_change_set.rs

```diff
@@ -0,0 +1,507 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    aggregator_extension::{
+        validate_history, AggregatorID, AggregatorState, DeltaHistory, SpeculativeStartValue,
+    },
+    bounded_math::{abort_error, addition_deltavalue, ok_overflow, ok_underflow},
+};
+use aptos_state_view::StateView;
+use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_core_types::vm_status::{StatusCode, VMStatus};
+
+/// When merging aggregator changes of two transactions,
+/// unable to merge the histories of the transactions.
+pub(crate) const EMERGE_HISTORIES: u64 = 0x02_0008;
+
+/// Represents a single aggregator change.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct AggregatorChange {
+    pub max_value: u128,
+    pub state: AggregatorState,
+    // `base_aggregator` is None for Aggregators.
+    // `base_aggregator` is Some(id) for AggregatorSnapshots.
+    pub base_aggregator: Option<AggregatorID>,
+}
+
+impl AggregatorChange {
+    /// TODO: Implement this.
+    pub fn materialize_aggregator(
+        self,
+        _state_view: &dyn StateView,
+        _aggregator_id: &AggregatorID,
+    ) -> anyhow::Result<AggregatorChange, VMStatus> {
+        Ok(self)
+    }
+
+    /// Returns the result of applying the AggregatorState to the base value
+    /// Returns error if postcondition is not satisfied.
+    pub fn apply_aggregator_change_to(&self, base: u128) -> PartialVMResult<u128> {
+        match self.state {
+            AggregatorState::Data { value } => Ok(value),
+            AggregatorState::Delta { delta, .. } => {
+                // First, validate if the current delta operation can be applied to the base.
+                validate_history(base, self.max_value, &self.state)?;
+                Ok(addition_deltavalue(base, delta, self.max_value)?)
+            },
+        }
+    }
+
+    /// Applies self on top of previous delta, merging them together. Note
+    /// that the strict ordering here is crucial for catching overflows
+    /// correctly.
+    ///
+    /// TODO: What happens if the base aggregator is not none?
+    pub fn merge_with_previous_aggregator_change(
+        &mut self,
+        previous_change: AggregatorChange,
+    ) -> PartialVMResult<()> {
+        assert_eq!(
+            self.max_value, previous_change.max_value,
+            "Cannot merge aggregator changes with different max_values",
+        );
+        // When the previous aggregator change is a snapshot, we cannot merge it with the current.
+        if previous_change.base_aggregator.is_some() {
```

Comment:
We can factor out `base_aggregator.is_some()` as `is_snapshot()`

Reply:
in the following diff, snapshot might also not have a base aggregator - for example if we call aggregator_v2::create_snapshot(value)